test/testkey/bucket.go

```diff
@@ -0,0 +1,116 @@
+package testkey
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+
+	"github.com/spiffe/spire/pkg/common/pemutil"
+)
+
+var (
+	packageDir string
+)
+
+func init() {
+	packageDir = initPackageDir()
+}
+
+func initPackageDir() string {
+	_, file, _, ok := runtime.Caller(0)
+	if !ok {
+		panic("unable to obtain caller information")
+	}
+	return filepath.Dir(file)
+}
+
+type keyType[K crypto.Signer] interface {
+	Path() string
+	GenerateKey() (K, error)
+}
+
+type bucket[KT keyType[K], K crypto.Signer] struct {
+	kt KT
+
+	mtx  sync.Mutex
+	keys []K
+}
+
+func (b *bucket[KT, K]) At(n int) (key K, err error) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	if err := b.load(); err != nil {
+		return key, err
+	}
+
+	switch {
+	case n > len(b.keys):
+		return key, errors.New("cannot ask for key beyond the end")
+	case n < len(b.keys):
+		return b.keys[n], nil
+	default:
+		key, err = b.kt.GenerateKey()
+		if err != nil {
+			return key, err
+		}
+		b.keys = append(b.keys, key)
+		if err := b.save(); err != nil {
+			return key, err
+		}
+		return key, nil
+	}
+}
+
+func (b *bucket[KT, K]) load() error {
+	if b.keys != nil {
+		return nil
+	}
+
+	blocks, err := pemutil.LoadBlocks(b.path())
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil
+		}
+		return err
+	}
+
+	var keys []K
+	for _, block := range blocks {
+		key, ok := block.Object.(K)
+		if !ok {
+			return fmt.Errorf("expected %T; got %T", key, block.Object)
+		}
+		keys = append(keys, key)
+	}
+
+	b.keys = keys
+	return nil
+}
+
+func (b *bucket[KT, K]) save() error {
+	buf := new(bytes.Buffer)
```

Comment:
```suggestion
	var buf bytes.Buffer
```
nit: the zero value of `bytes.Buffer` is ready to use

Reply:
I go back and forth on which pattern is more ergonomic. When using a zero valued var, you end up having to do `&buf` everywhere. Compiler escape analysis likely makes either form identical in terms of allocations. Anywho, happy to change.

Reply:
Yeah, they are basically equivalent. I tend to see the zero valued var usage more often and in Go documentation examples, so I generally lean towards that form.